modules/negotiate.go

```diff
@@ -0,0 +1,260 @@
+		NetAddress         NetAddress        `json:"netaddress"`
+		RemainingStorage   uint64            `json:"remainingstorage"`
+		SectorSize         uint64            `json:"sectorsize"`
+		TotalStorage       uint64            `json:"totalstorage"`
+		UnlockHash         types.UnlockHash  `json:"unlockhash"`
+		WindowSize         types.BlockHeight `json:"windowsize"`
+
+		// Collateral is the amount of collateral that the host will put up for
+		// storage in 'bytes per block', as an assurance to the renter that the
+		// host really is committed to keeping the file. But, because the file
+		// contract is created with no data available, this does leave the host
+		// exposed to an attack by a wealthy renter whereby the renter causes
+		// the host to lockup in-advance a bunch of funds that the renter then
+		// never uses, meaning the host will not have collateral for other
+		// clients.
+		//
+		// To mitigate the effects of this attack, the host has a collateral
+		// fraction and a max collateral. CollateralFraction is a number that
+		// gets divided by 1e6 and then represents the ratio of funds that the
+		// host is willing to put into the contract relative to the number of
+		// funds that the renter put into the contract. For example, if
+		// 'CollateralFraction' is set to 1e6 and the renter adds 1 siacoin of
+		// funding to the file contract, the host will also add 1 siacoin of
+		// funding to the contract. if 'CollateralFraction' is set to 2e6, the
+		// host would add 2 siacoins of funding to the contract.
+		//
+		// MaxCollateral indicates the maximum number of coins that a host is
+		// willing to put into a file contract.
+		Collateral         types.Currency `json:"collateral"`
+		CollateralFraction types.Currency `json:"collateralfraction"`
+		MaxCollateral      types.Currency `json:"maxcollateral"`
+
+		ContractPrice          types.Currency `json:"contractprice"`
+		DownloadBandwidthPrice types.Currency `json:"downloadbandwidthprice"`
+		StoragePrice           types.Currency `json:"storageprice"`
+		UploadBandwidthPrice   types.Currency `json:"uploadbandwidthprice"`
+
+		// Because the host has a public key, and settings are signed, and
+		// because settings may be MITM'd, settings need a revision number so
+		// that a renter can compare multiple sets of settings and determine
+		// which is the most recent.
+		RevisionNumber uint64 `json:"revisionnumber"`
+		Version        string `json:"version"`
+	}
+
+	// A RevisionAction is a description of an edit to be performed on a
+	// contract.
+	RevisionAction struct {
+		Type        types.Specifier
+		SectorIndex uint64
+		Offset      uint64
+		Data        []byte
+	}
+)
+
+// CreateAnnouncement will take a host announcement and encode it, returning
+// the exact []byte that should be added to the arbitrary data of a
+// transaction.
+func CreateAnnouncement(addr NetAddress, pk types.SiaPublicKey, sk crypto.SecretKey) (signedAnnouncement []byte, err error) {
```

Comment:
The `HostAnnouncement` type is used by the hostdb, but seems like it's not used anywhere else. I'm not sure that it does need to be exported, except that it is going into the blockchain in a specific way, so it's probably nice that the type is appearing in our godoc. I don't think it would be too hard to pull the host announcement out of the hostdb, seems like it's just passthrough anyway.

I'm inclined to leave it for now. I was hoping to leave things open so that we could switch to the bitcoin curve + schnorr, and the libraries to do that are now available. We should talk about that more at some point.

I do wish we didn't have to mix those types.